### PR TITLE
BFS Shortest Path

### LINKED ISSUE #469: BFS Shortest Path
Hi,
Do you think a BFS shortest path implementation would be useful? 
Cause currently one is bound to use Dijkstra/others also for unweighted graphs and BFS would perform much better in this case.

Regards,
Assaf.

### PR DESCRIPTION
**Pull request by ksskreddy**
- Implemented BFSShortestPath.
- Implemented UnitTest Cases
- This hopefully fixes #469

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

**Comment by jkinable** [comment]
Can you add this to the shortest path performance test and show that it outperforms some of the exiting methods?

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BFSShortestPath.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    {
+        if (!graph.containsVertex(source)) {
+            throw new IllegalArgumentException(GRAPH_MUST_CONTAIN_THE_SOURCE_VERTEX);
+        }
+
+        /*
+         * Initialize distance and predecessor.
+         */
+        int n = graph.vertexSet().size();
+        Map<V, Double> distance = new HashMap<>();
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BFSShortestPath.java]: No need to use several hashtables as we are doing for Dijkstra. The visited can be completely removed. You should be able to implement this directly using only `distanceAndPredecessorMap`.

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BFSShortestPath.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        while(!queue.isEmpty())
+        {
+            V v = queue.poll();
+            for (E e : graph.outgoingEdgesOf(v)) {
+                V u = Graphs.getOppositeVertex(graph, e, v);
+                if(!visited.containsKey(u))
+                {
+                    visited.put(u, true);
+                    queue.add(u);
+                    double newDist = distance.get(v) + graph.getEdgeWeight(e);
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BFSShortestPath.java]: The graph maybe weighted in which case you are not going to count the number of edges. Use `1` instead of `graph.getEdgeWeight(e)`.

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BFSShortestPath.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        Map<V, E> pred = new HashMap<>();
+        for (V v : graph.vertexSet()) {
+            distance.put(v, Double.POSITIVE_INFINITY);
+        }
+        distance.put(source, 0d);
+        
+        /*
+         * Declaring queue and visited array 
+         */
+        Queue<V> queue = new LinkedList<>();
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BFSShortestPath.java]: Use an `ArrayDeque` here as it much more (cache) friendly with modern processors.

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BFSShortestPath.java (hunk 2 above)
Comment (ksskreddy) [reply jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BFSShortestPath.java]: This algo works with weighted graphs with equal weighted edges.so i used it

**Comment by ksskreddy** [comment]
I have done the requested changes
Here are the results of Performance Tests

------
All-Pairs Shortest Paths Benchmark
Using G(n,p) random graph with n = 250, p = 0.3
Warmup phase 5 executions
Averaging results over 10 executions

| Shortest Path Algorithm                    	| Setup Time  	| ExecutionTime  	|
|--------------------------------------------	|-------------	|----------------	|
| Dijkstra with ClosestFirstIterator         	| 13.400 (ms) 	| 38320.300 (ms) 	|
| Dijkstra                                   	| 10.400 (ms) 	| 20328.400 (ms) 	|
| A* no heuristic                            	| 10.300 (ms) 	| 29235.100 (ms) 	|
| A* with ALT heuristic (1 random landmarks) 	| 9.500 (ms)  	| 34539.900 (ms) 	|
| A* with ALT heuristic (5 random landmarks) 	| 10.900 (ms) 	| 35862.800 (ms) 	|
| Bidirectional Dijkstra                     	| 8.800 (ms)  	| 8479.500 (ms)  	|
| BFSShortestPath                            	| 10.100 (ms) 	| 23507.400 (ms) 	|

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BFSShortestPath.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+import org.jgrapht.*;
+import org.jgrapht.alg.interfaces.ShortestPathAlgorithm.*;
+import org.jgrapht.alg.util.*;
+import org.jgrapht.graph.*;
+
+/**
+ * The BFS Shortest Path algorithm.
+ *
+ * <p>
+ * An implementation of <a href="https://www.geeksforgeeks.org/shortest-path-unweighted-graph/">BFS
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BFSShortestPath.java]: Replace the url with https://en.wikipedia.org/wiki/Breadth-first_search.

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BFSShortestPath.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    public BFSShortestPath(Graph<V, E> graph)
+    {
+        super(graph);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    @SuppressWarnings("unchecked")
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BFSShortestPath.java]: Why is this needed? Remove it.

**Comment by ksskreddy** [comment]
Any more improvements needed?

File: jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/BFSShortestPathTest.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+import java.util.*;
+
+import org.jgrapht.*;
+import org.jgrapht.alg.interfaces.ShortestPathAlgorithm.*;
+import org.jgrapht.graph.*;
+
+public class BFSShortestPathTest
+    extends
+    UnWeightedShortestPathTestCase
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/BFSShortestPathTest.java]: No need to use inheritance here.

File: jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/BFSShortestPathTest.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+{
+
+    @Override
+    protected List<DefaultEdge> findPathBetween(
+        Graph<String, DefaultEdge> g, String src, String dest)
+    {
+        SingleSourcePaths<String, DefaultEdge> tree;
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/BFSShortestPathTest.java]: We already have a static method `findPathBetween`. You can use that one.

File: jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/BFSShortestPathTest.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        tree = new BFSShortestPath<>(g).getPaths(src);
+        GraphPath<String, DefaultEdge> path = tree.getPath(dest);
+        if(path!=null)
+            return tree.getPath(dest).getEdgeList();
+        else 
+            return new LinkedList<DefaultEdge> ();
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/BFSShortestPathTest.java]: No need to hide the null values. Use `assertNull` in your checks.

File: jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/BFSShortestPathTest.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    protected SingleSourcePaths<String, DefaultEdge> getPathsFrom(
+        Graph<String, DefaultEdge> g, String src)
+    {
+        
+        return new BFSShortestPath<>(g).getPaths(src);
+    }
+
+    @Override
+    protected List<DefaultEdge> findPathTo(SingleSourcePaths<String, DefaultEdge> tree, String dest)
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/BFSShortestPathTest.java]: Similar comments here. No need for these helper methods. Just use `assertNull` where appropriate.

File: jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/UnWeightedShortestPathTestCase.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    public void testPathBetween()
+    {
+        List<DefaultEdge> path;
+        Graph<String, DefaultEdge> g = create();
+
+        path = findPathBetween(g, V1, V2);
+        assertEquals(Arrays.asList(new DefaultEdge[] { e12 }), path);
+
+        path = findPathBetween(g, V1, V4);
+        assertEquals(Arrays.asList(new DefaultEdge[] { e12, e24 }), path);
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/UnWeightedShortestPathTestCase.java]: `Arrays.asList(e12,e24)`

File: jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/UnWeightedShortestPathTestCase.java [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+     * .
+     */
+    @Test
+    public void testPathBetween()
+    {
+        List<DefaultEdge> path;
+        Graph<String, DefaultEdge> g = create();
+
+        path = findPathBetween(g, V1, V2);
+        assertEquals(Arrays.asList(new DefaultEdge[] { e12 }), path);
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/UnWeightedShortestPathTestCase.java]: `Arrays.asList(e12)`

File: jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/UnWeightedShortestPathTestCase.java [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+        path = findPathBetween(g, V1, V4);
+        assertEquals(Arrays.asList(new DefaultEdge[] { e12, e24 }), path);
+
+        path = findPathBetween(g, V1, V5);
+        assertEquals(Arrays.asList(new DefaultEdge[] { e13, e35 }), path);
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/UnWeightedShortestPathTestCase.java]: Same here

File: jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/UnWeightedShortestPathTestCase.java [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+        path = findPathBetween(g, V4, V3);
+        assertEquals(Arrays.asList(new DefaultEdge[] { }), path);
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/UnWeightedShortestPathTestCase.java]: `assertNull`

File: jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/UnWeightedShortestPathTestCase.java [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+    }
+    
+    
+    /*
+     * To test all the paths from a single source
+     */
+    
+    @Test
+    public void testAllPaths()
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/UnWeightedShortestPathTestCase.java]: Same comments as above here.

Comment (ksskreddy) [reply jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/UnWeightedShortestPathTestCase.java]: Done the proposed  changes

**Comment by ksskreddy** [comment]
Done the proposed changes

**Comment by d-michail** [comment]
Thanks
